packages/repository/src/define-repository-class.ts

```diff
@@ -0,0 +1,156 @@
+// Copyright IBM Corp. 2020. All Rights Reserved.
+// Node module: @loopback/repository
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import {Constructor} from '@loopback/context';
+import assert from 'assert';
+import {Entity, Model} from './model';
+import {
+  CrudRepository,
+  DefaultCrudRepository,
+  DefaultKeyValueRepository,
+  EntityCrudRepository,
+  juggler,
+  KeyValueRepository,
+} from './repositories';
+
+/**
+ * Signature for CrudRepository classes
+ *
+ * @typeParam M - Model class
+ * @typeParam R - Repository class/interface
+ */
+export interface CrudRepositoryClass<
+  M extends Model,
+  R extends CrudRepository<M>
+> {
+  new (ds: juggler.DataSource): R;
+}
+
+/**
+ * Signature for EntityCrudRepository classes
+ *
+ * @typeParam E - An entity class
+ * @typeParam IdType - ID type for the entity
+ * @typeParam Relations - Relations for the entity
+ */
+export interface EntityCrudRepositoryClass<
+  E extends Entity,
+  IdType,
+  Relations extends object
+> extends CrudRepositoryClass<E, EntityCrudRepository<E, IdType, Relations>> {}
+
+/**
+ * Create (define) a CRUD repository class for the given model.
+ *
+ * @example
+ *
+ * ```ts
+ * const AddressRepository = defineCrudRepositoryClass(Address);
+ * ```
+ *
+ * @param modelClass - A model class such as `Address`.
+ *
+ * @typeParam M - Model class
+ * @typeParam R - Repository class/interface
+ */
+export function defineCrudRepositoryClass<
+  M extends Model,
+  R extends CrudRepository<M>
+>(
+  modelClass: typeof Model & {prototype: M},
+  baseRepositoryClass: Constructor<R>,
```

Comment:
I find this construct confusing.

First, you say that `R extends CrudRepository<M>`, meaning that `R` must provide a constructor accepting a single argument `ds: juggler.DataSource`.

Next, you say that `baseRepositoryClass` can be a class extending `R` and having arbitrary constructor.

Why are we pinning down the allowed constructor arguments in `R` then? AFAICT, we never rely on that list of known arguments because it's always replaced with the arbitrary list of args described by `Constructor`.

Can you please take a look at the approach we described for repository mixins in https://github.com/strongloop/loopback-next/pull/4723? I think it would be good to find a single solution that we can use consistently in all places.

Have you considered an API like shown in the following snippet?

```ts
export function defineCrudRepositoryClass<<
   M extends Model,
   R extends Constructor<CrudRepository<M>>
 >(
  modelClass: typeof Model & {prototype: M},
  baseRepositoryClass,
) {
  // ...
}
```

Considering how this factory function works, I think we should strongly consider providing strong type checks for the repository constructor arguments supported by the factory function and also for the class returned by the factory:

1. The base repository must accept `(ModelCtor, dataSource)` (and exactly in this order).
2. The class returned by the factory accepts a single `(ModelCtor)` argument.

In the future, we can allow additional constructor arguments (`(ModelCtor, dataSource, ...args)` on input and `(ModelCtor, ...args)` in output), but let's leave that out of scope of this PR - it's already too complex.

Reply:
Uh oh, after reading the second commit, I realized that I got confused and the following part of my comment is not true:

> First, you say that `R extends CrudRepository<M>`, meaning that R must provide a constructor accepting a single argument `ds: juggler.DataSource`.

Your code is saying that `R` can have arbitrary constructor.

Which is still a problem, because we need `R` to accept `(ModelCtor, dataSource)` (and exactly in this order).